src/Algebra/Graph.hs

```diff
@@ -1066,3 +1104,77 @@ sparsify graph = res
         m <- get
         put (m + 1)
         overlay <$> s `x` m <*> m `y` t
+
+{- Note [The rules of foldg]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The rules for foldg work like this.
+
+* Up to (but not including) phase 1, we use the "buildR/f" rule to
+  rewrite all saturated applications of f with its buildR/foldg
+  form, hoping for fusion to happen (through the "foldg/buildR" rule).
+
+  In phase 1 and 0, we switch off these rules, inline buildR, and
+  switch on the "graph/f" rule, which rewrites the "foldg/f"
+  thing back into plain function if needed.
+
+  It's important that these two rules aren't both active at once
+  (along with build's unfolding) else we'd get an infinite loop
+  in the rules. Hence the activation control below.
+
+* composeR and matchR are here to remember the original function after
+  applying a "builR/f" rule. These functions are higher-order functions
+  and therefore benefits from inlining in final phase.
+
+* The "fmapR/fmapR" rule optimises compositions of fmapR
+-}
+
+type Foldg a = forall b. b -> (a -> b) -> (b -> b -> b) -> (b -> b -> b) -> b
+
+buildR :: forall a. Foldg a -> Graph a
```

Comment:
I guess the `forall a.` part is redundant?

Reply:
Oops yes